akka-stream/src/main/scala/akka/stream/impl/fusing/FoldWithin.scala

```diff
@@ -0,0 +1,147 @@
+ * @param emitOnTimer decide whether the current aggregator can be emitted, invoked after each timer event
+ * @param harvest     this is invoked as soon as all conditions are met before emitting to next stage
+ *                    there can be extra time between harvest and next stage receiving the emitted output
+ *                    time sensitive operations can be added here, such as closing an output channel
+ */
+class FoldWith[In, Agg, Out](
+    seed: In => Agg,
+    aggregate: (Agg, In) => Agg,
+    emitOnAgg: Agg => Boolean,
+    harvest: Agg => Out,
+    emitOnTimer: Option[(Agg => Boolean, FiniteDuration)] = None,
+    pullOnStart: Boolean = true)
+    extends GraphStage[FlowShape[In, Out]] {
+
+  emitOnTimer.foreach {
+    case (_, interval) => require(interval.gteq(1.milli), s"timer(${interval.toCoarsest}) must not be smaller than 1ms")
+  }
+
+  val in: Inlet[In] = Inlet[In](s"${this.getClass.getName}.in")
+  val out: Outlet[Out] = Outlet[Out](s"${this.getClass.getName}.out")
+  override val shape: FlowShape[In, Out] = FlowShape(in, out)
+
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic =
+    new TimerGraphStageLogic(shape) with InHandler with OutHandler {
+
+      // mutable state to keep track of the aggregator status to coordinate the flow
+      // input/output handler callbacks are guaranteed to execute without concurrency
+      // https://doc.akka.io/docs/akka/current/stream/stream-customize.html#thread-safety-of-custom-operators
+      private[this] var aggregator: Agg = null.asInstanceOf[Agg]
+
+      override def preStart(): Unit = {
+        if (pullOnStart) pull(in)
+        emitOnTimer.foreach {
+          case (_, interval) => scheduleWithFixedDelay("FoldWithinIntervalTimer", interval, interval)
+        }
+      }
+
+      override protected def onTimer(timerKey: Any): Unit =
+        emitOnTimer.foreach {
+          case (isReadyOnTimer, _) => if (aggregator != null && isReadyOnTimer(aggregator)) harvestAndEmit()
+        }
+
+      override def onPush(): Unit = {
+        aggregateAndEmitIfReady()
+        if (isAvailable(out)) pull(in) // pull only if there is no backpressure from outlet
+      }
+
+      override def onUpstreamFinish(): Unit = {
+        harvestAndEmit()
+        completeStage()
+      }
+
+      override def onPull(): Unit = {
+        aggregateAndEmitIfReady()
+        if (!hasBeenPulled(in)) pull(in) // always pass through the pull to upstream
+      }
+
+      setHandlers(in, out, this)
+
+      private def harvestAndEmit(): Unit =
+        if (aggregator != null) {
+          emit(out, FoldWith.this.harvest(aggregator))
+          aggregator = null.asInstanceOf[Agg]
+        }
+
+      private def aggregateAndEmitIfReady(): Unit = if (isAvailable(in)) {
+        val input = grab(in)
+        aggregator = if (aggregator == null) seed(input) else aggregate(aggregator, input)
+        if (emitOnAgg(aggregator)) harvestAndEmit()
+      }
```

Comment:
upstream data can be available on the inlet during downstream initiated onPull, so it can grab data and potentially emit  right there.
It's true it can wait for the next onPush and simply just call pull(in) here.
In that case it does not even need to check isAvailable(in) since it's always true for onPush.
I like the further logical simplification.

Reply:
merged the 2 parameters, there is quite some change in the signature